### PR TITLE
fix: improve query rewrite and fix bugs.

### PR DESCRIPTION
**Pull request by RinChanNOWWW**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

### Aggregating index

- Allow aggregation function `AVG` in aggregating index.
- In the past, if one group by key is an expression, Databend will not use aggregating index. This PR fixed this.
- The output fields of `TableScan` will the same order as the table schema. (In the past, the fields were sorted by names).
- Fix the memory violation while merging aggregate state.

### Aggregation

- Drop temporary state every time to avoid memory leak.
- Sort the input scalars of aggregation by index.

### CODE REVIEW

**Review by ariesdevil — APPROVED** [review]

File: src/query/service/src/pipelines/processors/transforms/aggregator/transform_aggregate_partial.rs [hunk 1]
```diff
@@ -200,11 +200,19 @@ impl<Method: HashMethodBounds> TransformPartialAggregate<Method> {
...
-                function.merge(place.next(offset), self.temp_place)?;
+                function.deserialize(state_place, &mut raw_state)?;
+                function.merge(place.next(offset), state_place)?;
+                if function.need_manual_drop_state() {
+                    unsafe {
+                        // State may allocate memory out of the arena,
+                        // drop state to avoid memory leak.
+                        function.drop_state(state_place);
+                    }
+                    function.init_state(state_place);
```
Comment (sundy-li) [inline src/query/service/src/pipelines/processors/transforms/aggregator/transform_aggregate_partial.rs:214]: the state is override by `deserialize`

Comment (RinChanNOWWW) [reply src/query/service/src/pipelines/processors/transforms/aggregator/transform_aggregate_partial.rs:214]: `deserialize` will call `place.get::<State>` first. If we do this without `init_state` after `drop_state`, we will violate the memory.

**Comment by Dousir9** [comment]
Need to rerun ci-benchmark.

File: src/query/ee/tests/it/aggregating_index/index_scan.rs [hunk 2]
```diff
@@ -56,6 +56,12 @@ async fn test_index_scan_with_count() -> Result<()> {
     test_index_scan_with_count_impl("native").await
 }
 
+#[tokio::test(flavor = "multi_thread")]
+async fn test_index_scan_agg_args_are_expression() -> Result<()> {
+    test_index_scan_agg_args_are_expression_impl("parquet").await?;
+    test_index_scan_agg_args_are_expression_impl("native").await
+}
```
Comment (RinChanNOWWW) [inline src/query/ee/tests/it/aggregating_index/index_scan.rs:63]: A comment to prevent from being merged.

**Comment by BohuTANG** [comment]
> ## ClickBench Report
> * **hits**: https://repo.databend.rs/benchmark/clickbench/pr/12140/5617431933/hits.html
> * **tpch**: https://repo.databend.rs/benchmark/clickbench/pr/12140/5617431933/tpch.html

The report is broken :\

**Pull request merged**